File: frontend/apps/android/app/src/main/java/app/pantopus/android/ui/screens/shared/content_detail/bodies/BodyReactionsBody.kt
@file:Suppress("MagicNumber", "PackageNaming", "LongParameterList", "LongMethod", "TooManyFunctions")

package app.pantopus.android.ui.screens.shared.content_detail.bodies

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.aspectRatio
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.sizeIn
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.OutlinedTextFieldDefaults
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import app.pantopus.android.ui.components.AvatarWithIdentityRing
import app.pantopus.android.ui.components.IdentityPillar
import app.pantopus.android.ui.theme.PantopusColors
import app.pantopus.android.ui.theme.PantopusIcon
import app.pantopus.android.ui.theme.PantopusIconImage
import app.pantopus.android.ui.theme.Radii
import app.pantopus.android.ui.theme.Spacing
import coil.compose.SubcomposeAsyncImage

/** One row in the flattened comment thread. */
data class PostCommentRow(
    val id: String,
    val authorName: String,
    val authorAvatarUrl: String?,
    val authorIdentity: IdentityPillar,
    val body: String,
    val timestamp: String,
    /** 0 for top-level, 1 for nested. Deeper threads collapse to 1. */
    val indentLevel: Int,
    val authorUserId: String?,
)

/** Reaction-bar counts + the user's currently-selected reaction (if any). */
data class PostReactionCounts(
    val helpful: Int = 0,
    val heart: Int = 0,
    val going: Int = 0,
    val userReaction: app.pantopus.android.data.api.models.posts.PostReactionKind? = null,
)

/**
 * Pulse post body — text + media + reactions + comment composer + flat
 * thread. Pure render surface; all state lives in the host VM.
 */
@Composable
fun BodyReactionsBody(
    body: String,
    mediaUrls: List<String>,
    reactions: PostReactionCounts,
    onReactionTap: (app.pantopus.android.data.api.models.posts.PostReactionKind) -> Unit,
    composerAvatarUrl: String?,
    composerAvatarName: String,
    composerText: String,
    onComposerTextChange: (String) -> Unit,
    isSending: Boolean,
    onSendTap: () -> Unit,
    comments: List<PostCommentRow>,
    hiddenReplyCount: Int = 0,
    onShowMoreReplies: (() -> Unit)? = null,
    onCommentAvatarTap: (String) -> Unit = {},
) {
    Column(
        modifier = Modifier.fillMaxWidth(),
        verticalArrangement = Arrangement.spacedBy(Spacing.s4),
    ) {
        if (body.isNotEmpty()) {
            Text(
                text = body,
                fontSize = 15.sp,
                color = PantopusColors.appText,
                lineHeight = 22.sp,
                modifier = Modifier.padding(horizontal = Spacing.s4),
            )
        }
        if (mediaUrls.isNotEmpty()) {
            PostMediaGrid(
                urls = mediaUrls,
                modifier = Modifier.padding(horizontal = Spacing.s4),
            )
        }
        ReactionsBar(
            counts = reactions,
            onTap = onReactionTap,
            modifier = Modifier.padding(horizontal = Spacing.s4),
        )
        CommentComposer(
            avatarName = composerAvatarName,
            avatarUrl = composerAvatarUrl,
            text = composerText,
            onTextChange = onComposerTextChange,
            isSending = isSending,
            onSend = onSendTap,
            modifier = Modifier.padding(horizontal = Spacing.s4),
        )
        if (comments.isNotEmpty()) {
            Column(
                modifier = Modifier.padding(horizontal = Spacing.s4),
                verticalArrangement = Arrangement.spacedBy(Spacing.s3),
            ) {
                comments.forEach { row ->
                    CommentRow(
                        row = row,
                        onAvatarTap = {
                            row.authorUserId?.let(onCommentAvatarTap)
                        },
                    )
                }
                if (hiddenReplyCount > 0 && onShowMoreReplies != null) {
                    Text(
                        text =
                            "View $hiddenReplyCount more " +
                                if (hiddenReplyCount == 1) "reply" else "replies",
                        fontSize = 14.sp,
                        fontWeight = FontWeight.SemiBold,
                        color = PantopusColors.primary600,
                        modifier =
                            Modifier
                                .padding(start = (28 + 12).dp)
                                .heightIn(min = 44.dp)
                                .clickable(onClick = onShowMoreReplies)
                                .semantics { contentDescription = "View $hiddenReplyCount more replies" },
                    )
                }
            }
        }
    }
}

@Composable
private fun PostMediaGrid(
    urls: List<String>,
    modifier: Modifier = Modifier,
) {
    when (urls.size) {
        0 -> Unit
        1 -> {
            MediaTile(
                url = urls[0],
                modifier =
                    modifier
                        .fillMaxWidth()
                        .aspectRatio(16f / 9f)
                        .clip(RoundedCornerShape(Radii.lg)),
            )
        }
        2 -> {
            Row(
                modifier = modifier.fillMaxWidth().height(160.dp).clip(RoundedCornerShape(Radii.lg)),
                horizontalArrangement = Arrangement.spacedBy(Spacing.s2),
            ) {
                MediaTile(url = urls[0], modifier = Modifier.weight(1f).fillMaxWidth())
                MediaTile(url = urls[1], modifier = Modifier.weight(1f).fillMaxWidth())
            }
        }
        3 -> {
            Row(
                modifier = modifier.fillMaxWidth().height(200.dp).clip(RoundedCornerShape(Radii.lg)),
                horizontalArrangement = Arrangement.spacedBy(Spacing.s2),
            ) {
                MediaTile(url = urls[0], modifier = Modifier.weight(1f).fillMaxWidth())
                Column(
                    modifier = Modifier.weight(1f).fillMaxWidth(),
                    verticalArrangement = Arrangement.spacedBy(Spacing.s2),
                ) {
                    MediaTile(url = urls[1], modifier = Modifier.weight(1f).fillMaxWidth())
                    MediaTile(url = urls[2], modifier = Modifier.weight(1f).fillMaxWidth())
                }
            }
        }
        else -> {
            // 4+ → 2x2 of the first 4.
            Column(
                modifier = modifier.fillMaxWidth().clip(RoundedCornerShape(Radii.lg)),
                verticalArrangement = Arrangement.spacedBy(Spacing.s2),
            ) {
                Row(horizontalArrangement = Arrangement.spacedBy(Spacing.s2)) {
                    MediaTile(url = urls[0], modifier = Modifier.weight(1f).aspectRatio(1f))
                    MediaTile(url = urls[1], modifier = Modifier.weight(1f).aspectRatio(1f))
                }
                Row(horizontalArrangement = Arrangement.spacedBy(Spacing.s2)) {
                    MediaTile(url = urls[2], modifier = Modifier.weight(1f).aspectRatio(1f))
                    MediaTile(url = urls[3], modifier = Modifier.weight(1f).aspectRatio(1f))
                }
            }
        }
    }
}

@Composable
private fun MediaTile(
    url: String,
    modifier: Modifier = Modifier,
) {
    SubcomposeAsyncImage(
        model = url,
        contentDescription = null,
        modifier = modifier.background(PantopusColors.appSurfaceSunken),
        loading = {
            Box(modifier = Modifier.fillMaxWidth(), contentAlignment = Alignment.Center) {
                CircularProgressIndicator(strokeWidth = 2.dp, modifier = Modifier.size(20.dp))
            }
        },
        error = {
            Box(modifier = Modifier.fillMaxWidth(), contentAlignment = Alignment.Center) {
                PantopusIconImage(
                    icon = PantopusIcon.AlertCircle,
                    contentDescription = null,
                    size = 22.dp,
                    tint = PantopusColors.appTextMuted,
                )
            }
        },
    )
}

@Composable
private fun ReactionsBar(
    counts: PostReactionCounts,
    onTap: (app.pantopus.android.data.api.models.posts.PostReactionKind) -> Unit,
    modifier: Modifier = Modifier,
) {
    Row(
        modifier = modifier.fillMaxWidth(),
        horizontalArrangement = Arrangement.spacedBy(Spacing.s2),
    ) {
        ReactionPill(
            label = "Helpful",
            icon = PantopusIcon.ThumbsUp,
            count = counts.helpful,
            isSelected = counts.userReaction == app.pantopus.android.data.api.models.posts.PostReactionKind.Helpful,
            onTap = { onTap(app.pantopus.android.data.api.models.posts.PostReactionKind.Helpful) },
            accessibilityLabel = "Helpful, ${counts.helpful}",
        )
        ReactionPill(
            label = "Heart",
            icon = PantopusIcon.Heart,
            count = counts.heart,
            isSelected = counts.userReaction == app.pantopus.android.data.api.models.posts.PostReactionKind.Heart,
            onTap = { onTap(app.pantopus.android.data.api.models.posts.PostReactionKind.Heart) },
            accessibilityLabel = "Loved, ${counts.heart}",
        )
        ReactionPill(
            label = "Going",
            icon = PantopusIcon.Check,
            count = counts.going,
            isSelected = counts.userReaction == app.pantopus.android.data.api.models.posts.PostReactionKind.Going,
            onTap = { onTap(app.pantopus.android.data.api.models.posts.PostReactionKind.Going) },
            accessibilityLabel = "Going, ${counts.going}",
        )
    }
}

@Composable
private fun ReactionPill(
    label: String,
    icon: PantopusIcon,
    count: Int,
    isSelected: Boolean,
    onTap: () -> Unit,
    accessibilityLabel: String,
) {
    val background = if (isSelected) PantopusColors.primary600 else PantopusColors.appSurfaceSunken
    val foreground = if (isSelected) PantopusColors.appTextInverse else PantopusColors.appText
    Box(
        modifier =
            Modifier
                .clip(RoundedCornerShape(Radii.pill))
                .background(background)
                .sizeIn(minHeight = 44.dp)
                .clickable(onClick = onTap)
                .padding(horizontal = Spacing.s3, vertical = 6.dp)
                .semantics { contentDescription = accessibilityLabel },
        contentAlignment = Alignment.Center,
    ) {
        Row(
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.spacedBy(Spacing.s1),
        ) {
            PantopusIconImage(icon = icon, contentDescription = null, size = 14.dp, tint = foreground)
            Text(
                text = label,
                fontSize = 12.sp,
                fontWeight = FontWeight.SemiBold,
                color = foreground,
            )
            Text(
                text = "$count",
                fontSize = 12.sp,
                color = foreground.copy(alpha = 0.85f),
            )
        }
    }
}

@Composable
private fun CommentComposer(
    avatarName: String,
    avatarUrl: String?,
    text: String,
    onTextChange: (String) -> Unit,
    isSending: Boolean,
    onSend: () -> Unit,
    modifier: Modifier = Modifier,
) {
    Row(
        modifier = modifier.fillMaxWidth(),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(Spacing.s2),
    ) {
        AvatarWithIdentityRing(
            name = avatarName,
            imageUrl = avatarUrl,
            identity = IdentityPillar.Personal,
            ringProgress = 1f,
            size = 28.dp,
        )
        OutlinedTextField(
            value = text,
            onValueChange = onTextChange,
            placeholder = { Text("Add a comment", fontSize = 14.sp) },
            singleLine = true,
            shape = RoundedCornerShape(Radii.pill),
            textStyle = TextStyle(fontSize = 14.sp),
            colors =
                OutlinedTextFieldDefaults.colors(
                    focusedBorderColor = PantopusColors.appBorder,
                    unfocusedBorderColor = PantopusColors.appBorder,
                    focusedContainerColor = PantopusColors.appSurfaceSunken,
                    unfocusedContainerColor = PantopusColors.appSurfaceSunken,
                ),
            modifier =
                Modifier
                    .weight(1f)
                    .height(44.dp)
                    .semantics { contentDescription = "Add a comment" },
        )
        val isEnabled = text.trim().isNotEmpty() && !isSending
        Box(
            modifier =
                Modifier
                    .size(44.dp)
                    .clip(CircleShape)
                    .background(if (isEnabled) PantopusColors.primary100 else Color.Transparent)
                    .clickable(enabled = isEnabled, onClick = onSend)
                    .semantics { contentDescription = "Send comment" },
            contentAlignment = Alignment.Center,
        ) {
            if (isSending) {
                CircularProgressIndicator(strokeWidth = 2.dp, modifier = Modifier.size(20.dp))
            } else {
                PantopusIconImage(
                    icon = PantopusIcon.Send,
                    contentDescription = null,
                    size = 18.dp,
                    tint = if (isEnabled) PantopusColors.primary600 else PantopusColors.appTextMuted,
                )
            }
        }
    }
}

@Composable
private fun CommentRow(
    row: PostCommentRow,
    onAvatarTap: () -> Unit,
) {
    Row(
        modifier = Modifier.fillMaxWidth().padding(vertical = 2.dp),
        verticalAlignment = Alignment.Top,
        horizontalArrangement = Arrangement.spacedBy(Spacing.s2),
    ) {
        if (row.indentLevel > 0) {
            Spacer(Modifier.width((row.indentLevel * 28).dp))
        }
        Box(
            modifier =
                Modifier
                    .sizeIn(minWidth = 28.dp, minHeight = 28.dp)
                    .clickable(onClick = onAvatarTap)
                    .semantics { contentDescription = "Open ${row.authorName}'s profile" },
        ) {
            AvatarWithIdentityRing(
                name = row.authorName,
                imageUrl = row.authorAvatarUrl,
                identity = row.authorIdentity,
                ringProgress = 1f,
                size = 28.dp,
            )
        }
        Column(modifier = Modifier.weight(1f)) {
            Row(
                verticalAlignment = Alignment.CenterVertically,
                horizontalArrangement = Arrangement.spacedBy(Spacing.s1),
            ) {
                Text(
                    text = row.authorName,
                    fontSize = 12.sp,
                    fontWeight = FontWeight.SemiBold,
                    color = PantopusColors.appText,
                )
                Text("·", fontSize = 11.sp, color = PantopusColors.appTextMuted)
                Text(
                    text = row.timestamp,
                    fontSize = 11.sp,
                    color = PantopusColors.appTextSecondary,
                )
            }
            Text(
                text = row.body,
                fontSize = 12.sp,
                color = PantopusColors.appTextStrong,
                lineHeight = 16.sp,
            )
        }
    }
}

File: frontend/apps/ios/Pantopus/Features/Shared/ContentDetail/Bodies/BodyReactionsBody.swift
//
//  BodyReactionsBody.swift
//  Pantopus
//
//  `body_reactions` slot for the Pulse post detail. Body copy → media
//  grid → reactions bar → inline composer → flattened comment thread.
//
// swiftlint:disable multiple_closures_with_trailing_closure

import SwiftUI

/// View-model surface for a single comment row.
public struct PostCommentRow: Sendable, Identifiable, Hashable {
    public let id: String
    public let authorName: String
    public let authorAvatarURL: URL?
    public let authorIdentity: IdentityPillar
    public let body: String
    public let timestamp: String
    public let reactionCount: Int
    public let userReacted: Bool
    /// 0 for top-level, 1 for nested. We collapse deeper threads to 1.
    public let indentLevel: Int
    public let authorUserId: String?

    public init(
        id: String,
        authorName: String,
        authorAvatarURL: URL?,
        authorIdentity: IdentityPillar,
        body: String,
        timestamp: String,
        reactionCount: Int = 0,
        userReacted: Bool = false,
        indentLevel: Int,
        authorUserId: String?
    ) {
        self.id = id
        self.authorName = authorName
        self.authorAvatarURL = authorAvatarURL
        self.authorIdentity = authorIdentity
        self.body = body
        self.timestamp = timestamp
        self.reactionCount = reactionCount
        self.userReacted = userReacted
        self.indentLevel = indentLevel
        self.authorUserId = authorUserId
    }
}

/// Quick-reply prompt rendered in the empty thread state.
public struct PostQuickReplyPrompt: Sendable, Identifiable, Hashable {
    public let id: String
    public let label: String
    public let icon: PantopusIcon

    public init(label: String, icon: PantopusIcon) {
        id = label
        self.label = label
        self.icon = icon
    }
}

/// Counts displayed under each reaction pill, plus which (if any) the
/// signed-in user has currently selected.
public struct PostReactionCounts: Sendable, Hashable {
    public var helpful: Int
    public var heart: Int
    public var going: Int
    public var userReaction: PostReactionKind?

    public init(helpful: Int = 0, heart: Int = 0, going: Int = 0, userReaction: PostReactionKind? = nil) {
        self.helpful = helpful
        self.heart = heart
        self.going = going
        self.userReaction = userReaction
    }
}

/// Pulse post body — text + media + reactions + comments. Pure render
/// surface; all state lives in the host view-model.
@MainActor
public struct BodyReactionsBody: View {
    private let bodyText: String
    private let mediaURLs: [URL]
    private let intent: PostIntent
    private let reactions: PostReactionCounts
    private let onReactionTap: @MainActor (PostReactionKind) -> Void
    private let composerAvatarURL: URL?
    private let composerAvatarName: String
    @Binding private var composerText: String
    private let isSending: Bool
    private let onSendTap: @MainActor () -> Void
    private let comments: [PostCommentRow]
    private let hiddenReplyCount: Int
    private let onShowMoreReplies: (@MainActor () -> Void)?
    private let onCommentAvatarTap: @MainActor (String) -> Void

    public init(
        body: String,
        mediaURLs: [URL],
        intent: PostIntent = .share,
        reactions: PostReactionCounts,
        onReactionTap: @escaping @MainActor (PostReactionKind) -> Void,
        composerAvatarURL: URL?,
        composerAvatarName: String,
        composerText: Binding<String>,
        isSending: Bool,
        onSendTap: @escaping @MainActor () -> Void,
        comments: [PostCommentRow],
        hiddenReplyCount: Int = 0,
        onShowMoreReplies: (@MainActor () -> Void)? = nil,
        onCommentAvatarTap: @escaping @MainActor (String) -> Void = { _ in }
    ) {
        bodyText = body
        self.mediaURLs = mediaURLs
        self.intent = intent
        self.reactions = reactions
        self.onReactionTap = onReactionTap
        self.composerAvatarURL = composerAvatarURL
        self.composerAvatarName = composerAvatarName
        _composerText = composerText
        self.isSending = isSending
        self.onSendTap = onSendTap
        self.comments = comments
        self.hiddenReplyCount = hiddenReplyCount
        self.onShowMoreReplies = onShowMoreReplies
        self.onCommentAvatarTap = onCommentAvatarTap
    }

    public var body: some View {
        VStack(alignment: .leading, spacing: Spacing.s4) {
            if !bodyText.isEmpty {
                Text(bodyText)
                    .font(.system(size: 15, weight: .regular))
                    .foregroundStyle(Theme.Color.appText)
                    .lineSpacing(7) // 15 + 7 ≈ 22pt line height
                    .padding(.horizontal, Spacing.s4)
                    .accessibilityLabel(bodyText)
            }

            if !mediaURLs.isEmpty {
                PostMediaGrid(urls: mediaURLs)
                    .padding(.horizontal, Spacing.s4)
            }

            ReactionsBar(
                counts: reactions,
                commentCount: visibleCommentCount,
                commentsAreFresh: comments.isEmpty,
                onTap: onReactionTap
            )
            .padding(.horizontal, Spacing.s4)

            Rectangle()
                .fill(Theme.Color.appBorder)
                .frame(height: 1)
                .padding(.horizontal, Spacing.s4)

            CommentComposer(
                avatarName: composerAvatarName,
                avatarURL: composerAvatarURL,
                text: $composerText,
                placeholder: comments.isEmpty ? "Be the first to reply..." : "Add a comment",
                isFocusedPresentation: comments.isEmpty,
                isSending: isSending,
                onSend: onSendTap
            )
            .padding(.horizontal, Spacing.s4)

            if !comments.isEmpty {
                VStack(alignment: .leading, spacing: Spacing.s3) {
                    ForEach(comments) { comment in
                        CommentRow(comment: comment) {
                            if let uid = comment.authorUserId { onCommentAvatarTap(uid) }
                        }
                    }
                    if hiddenReplyCount > 0, let onShowMoreReplies {
                        Button {
                            onShowMoreReplies()
                        } label: {
                            Text("View \(hiddenReplyCount) more \(hiddenReplyCount == 1 ? "reply" : "replies")")
                                .font(.system(size: PantopusTextStyle.small.size, weight: .semibold))
                                .foregroundStyle(Theme.Color.primary600)
                                .frame(minHeight: 44)
                        }
                        .buttonStyle(.plain)
                        .accessibilityLabel("View \(hiddenReplyCount) more replies")
                        .padding(.leading, 28 + Spacing.s3) // align with comment body
                    }
                }
                .padding(.horizontal, Spacing.s4)
            } else {
                EmptyThreadState(
                    intent: intent,
                    prompts: intent.quickReplyPrompts
                ) { prompt in
                    composerText = prompt.label
                }
                .padding(.horizontal, Spacing.s4)
            }
        }
    }

    private var visibleCommentCount: Int {
        comments.count + hiddenReplyCount
    }
}

// MARK: - Sub-views

private struct PostMediaGrid: View {
    let urls: [URL]

    var body: some View {
        switch urls.count {
        case 0: EmptyView()
        case 1:
            mediaTile(urls[0])
                .aspectRatio(16.0 / 9.0, contentMode: .fit)
                .clipShape(RoundedRectangle(cornerRadius: Radii.lg))
        case 2:
            HStack(spacing: Spacing.s2) {
                mediaTile(urls[0]).aspectRatio(1, contentMode: .fill)
                mediaTile(urls[1]).aspectRatio(1, contentMode: .fill)
            }
            .frame(height: 160)
            .clipShape(RoundedRectangle(cornerRadius: Radii.lg))
        case 3:
            HStack(spacing: Spacing.s2) {
                mediaTile(urls[0]).frame(maxWidth: .infinity)
                VStack(spacing: Spacing.s2) {
                    mediaTile(urls[1]).frame(maxWidth: .infinity)
                    mediaTile(urls[2]).frame(maxWidth: .infinity)
                }
            }
            .frame(height: 200)
            .clipShape(RoundedRectangle(cornerRadius: Radii.lg))
        default:
            LazyVGrid(
                columns: [GridItem(.flexible(), spacing: Spacing.s2), GridItem(.flexible(), spacing: Spacing.s2)],
                spacing: Spacing.s2
            ) {
                ForEach(Array(urls.prefix(4).enumerated()), id: \.offset) { _, url in
                    mediaTile(url).aspectRatio(1, contentMode: .fill)
                }
            }
            .clipShape(RoundedRectangle(cornerRadius: Radii.lg))
        }
    }

    private func mediaTile(_ url: URL) -> some View {
        AsyncImage(url: url) { phase in
            switch phase {
            case let .success(image):
                image.resizable().scaledToFill()
            case .failure:
                Color.gray.opacity(0.1)
                    .overlay(Icon(.alertCircle, size: 22, color: Theme.Color.appTextMuted))
            case .empty:
                Color.gray.opacity(0.08).overlay(ProgressView())
            @unknown default:
                Color.gray.opacity(0.08)
            }
        }
        .frame(maxWidth: .infinity)
        .clipped()
    }
}

private struct ReactionsBar: View {
    let counts: PostReactionCounts
    let commentCount: Int
    let commentsAreFresh: Bool
    let onTap: @MainActor (PostReactionKind) -> Void

    var body: some View {
        HStack(spacing: Spacing.s2) {
            ReactionPill(
                id: "heart",
                icon: .heart,
                count: counts.helpful,
                accessibilityLabel: "Heart reaction, \(counts.helpful)",
                isSelected: counts.userReaction == .helpful,
                selectedForeground: Theme.Color.error,
                selectedBackground: Theme.Color.errorBg
            ) { onTap(.helpful) }
            ReactionPill(
                id: "hand",
                icon: .hand,
                count: counts.heart,
                accessibilityLabel: "Raised hand reaction, \(counts.heart)",
                isSelected: counts.userReaction == .heart,
                onTap: nil
            )
            ReactionPill(
                id: "eye",
                icon: .eye,
                count: counts.going,
                accessibilityLabel: "Watching reaction, \(counts.going)",
                isSelected: counts.userReaction == .going,
                onTap: nil
            )
            Spacer()
            HStack(spacing: Spacing.s1) {
                Icon(.messageCircle, size: 13, color: commentsAreFresh ? Theme.Color.appTextMuted : Theme.Color.appTextSecondary)
                Text(commentSummary)
                    .font(.system(size: 11, weight: .regular))
                    .foregroundStyle(commentsAreFresh ? Theme.Color.appTextMuted : Theme.Color.appTextSecondary)
            }
            .accessibilityLabel(commentSummary)
        }
    }

    private var commentSummary: String {
        if commentsAreFresh {
            return "0 comments · just posted"
        }
        return "\(commentCount) \(commentCount == 1 ? "comment" : "comments")"
    }
}

private struct ReactionPill: View {
    let id: String
    let icon: PantopusIcon
    let count: Int
    let accessibilityLabel: String
    let isSelected: Bool
    var selectedForeground: Color = Theme.Color.primary700
    var selectedBackground: Color = Theme.Color.primary50
    /// `nil` renders the pill as display-only (no Button wrapper).
    let onTap: (@MainActor () -> Void)?

    var body: some View {
        if let onTap {
            Button(action: { onTap() }) { pillBody }
                .buttonStyle(.plain)
                .frame(minHeight: 44, alignment: .center)
                .accessibilityIdentifier("pulsePostDetail-reaction-\(id)")
                .accessibilityLabel(accessibilityLabel)
                .accessibilityAddTraits(isSelected ? [.isButton, .isSelected] : .isButton)
        } else {
            pillBody
                .accessibilityElement()
                .accessibilityIdentifier("pulsePostDetail-reaction-\(id)")
                .accessibilityLabel(accessibilityLabel)
        }
    }

    private var pillBody: some View {
        HStack(spacing: Spacing.s1) {
            Icon(icon, size: 14, color: foreground)
            Text("\(count)")
                .font(.system(size: PantopusTextStyle.caption.size, weight: .regular))
                .foregroundStyle(foreground)
        }
        .padding(.horizontal, Spacing.s3)
        .frame(height: 32)
        .background(background)
        .overlay {
            RoundedRectangle(cornerRadius: Radii.pill, style: .continuous)
                .stroke(border, lineWidth: 1)
        }
        .clipShape(RoundedRectangle(cornerRadius: Radii.pill))
    }

    private var foreground: Color {
        isSelected ? selectedForeground : Theme.Color.appTextSecondary
    }

    private var background: Color {
        isSelected ? selectedBackground : Theme.Color.appSurface
    }

    private var border: Color {
        isSelected ? selectedForeground.opacity(0.25) : Theme.Color.appBorder
    }
}

// MARK: - Preview

#Preview("Populated") {
    @Previewable @State var text = ""
    BodyReactionsBody(
        body: "Anyone know a reliable handyman for fixing a leaky faucet in Cambridge? Tried two on Angi already.",
        mediaURLs: [],
        reactions: PostReactionCounts(helpful: 7, heart: 2, going: 0, userReaction: .helpful),
        onReactionTap: { _ in },
        composerAvatarURL: nil,
        composerAvatarName: "You",
        composerText: $text,
        isSending: false,
        onSendTap: {},
        comments: [
            PostCommentRow(
                id: "1",
                authorName: "Maria Chen",
                authorAvatarURL: nil,
                authorIdentity: .personal,
                body: "I used Mike at Westside Plumbing last month. Solid.",
                timestamp: "3m ago",
                indentLevel: 0,
                authorUserId: "u1"
            ),
            PostCommentRow(
                id: "2",
                authorName: "Sam Lee",
                authorAvatarURL: nil,
                authorIdentity: .personal,
                body: "+1 to Mike. Fair pricing too.",
                timestamp: "2m ago",
                indentLevel: 1,
                authorUserId: "u2"
            )
        ],
        hiddenReplyCount: 4
    ) {}
        .background(Theme.Color.appBg)
}
